h2o-core/src/main/java/water/ExternalFrameWriter.java

```diff
@@ -0,0 +1,157 @@
+package water;
+
+import water.fvec.ChunkUtils;
+import water.fvec.NewChunk;
+
+import java.io.IOException;
+import java.nio.channels.SocketChannel;
+import java.sql.Timestamp;
+
+/**
+ * Use to create H2O Frames from non H2O environments
+ */
+public class ExternalFrameWriter {
+
+    private static final byte CREATE_NEW_CHUNK = 0;
+    private static final byte ADD_TO_CHUNK = 1;
+    private static final byte CLOSE_NEW_CHUNK = 2;
+
+    private static final byte TYPE_NUM = 0;
+    private static final byte TYPE_STR = 1;
+    private static final byte TYPE_NA = 2;
+
+    private AutoBuffer ab;
+    private SocketChannel channel;
+
+    public ExternalFrameWriter(SocketChannel channel){
+        // using default constructor, AutoBuffer is created with
+        // private property _read set to false, in order to satisfy call clearForWriting it has to be set to true
+        // which does the call of flipForReading method
+        ab = new AutoBuffer().flipForReading();
+        this.channel = channel;
+    }
+
+
+    public void createChunks(String keystr, byte[] vecTypes, int chunkId) throws IOException {
+        ab.clearForWriting(H2O.MAX_PRIORITY);
+        ab.put1(ExternalFrameHandler.INIT_BYTE);
+        ab.putInt(ExternalFrameHandler.CREATE_FRAME);
+        ab.putInt(CREATE_NEW_CHUNK);
+        ab.putStr(keystr);
+        ab.putA1(vecTypes);
+        ab.putInt(chunkId);
+        writeToChannel(ab, channel);
+    }
+
+    public void closeChunks() throws IOException{
+        ab.clearForWriting(H2O.MAX_PRIORITY);
+        ab.putInt(CLOSE_NEW_CHUNK);
+        writeToChannel(ab, channel);
+
+        AutoBuffer confirmAb = new AutoBuffer(channel, null);
+        // this needs to be here because confirmAb.getInt() forces this code to wait for result and
+        // all the previous work to be done on the recipient side. The assert around it is just additional, not
+        // so important check
+        assert(confirmAb.getInt() == ExternalFrameHandler.CONFIRM_WRITING_DONE);
+    }
+
+    private void setPutHeader(){
+        ab.clearForWriting(H2O.MAX_PRIORITY);
+        ab.putInt(ADD_TO_CHUNK);
+    }
+
+    public void put(int columnNum, double num) throws IOException{
+        setPutHeader();
```

Comment:
It needs change since right now the cost of sending a single value is too high because of included meta-data per sent value.